### ISSUE TITLE
[Question] should I dispose `when`?

### ISSUE THREAD

**Issue by subzero911** [mobxjs/mobx.dart#568]
It's stated in the docs:
> After the effect() is run, when automatically disposes itself.

It's a little misunderstanding.
What if the effect didn't run? Should I call dispose() manually in this case?
What if the effect runned and `when` is disposed, but I also call dispose() - will it cause error like `call dispose() on null`?

**Comment by pavanpodila** [comment]
`when` will only run the effect when the condition matches. Until then it is dormant. You can `dispose()` it prematurely by calling dispose yourself.

**Comment by subzero911** [comment]
It looks like this:

```dart
initState()
{
 disposer = when (...);
}

dispose()
{
 disposer();
}
```
What if the effect runs, and I call Navigator.pop() after it? dispose() will run, but the reaction is disposed already. Wouldn't that lead to error?

**Comment by Fleximex** [comment]
@subzero911 
I just tested it. I trigger the `when` reaction and then I increment a counter in a MobX store 3 seconds after I do `Navigator.pop` and dispose the reaction. The counter still increments even though the `when` reaction is disposed.

So when the reaction is triggered before it is disposed, the effect will keep running. The dispose call disposes the predicate check not the effect.

Of course, when you change something on the screen you just disposed (with `Navigator.pop`) it will throw an error. To check for that you can use `if (mounted) {}`. For other errors there is also a `onError` parameter you can use in the `when` reaction.

**Comment by subzero911** [comment]
So, is where a difference between?
```dart
_d = when(
      (_) => authStore.isAuth == false,
      onUserLogout,
    );
```
and
```dart
_d = reaction(
      (_) => authStore.isAuth,
      (isAuth) {
        if (!isAuth) onUserLogout();
      },
    );
```

**Comment by Fleximex** [comment]
That's correct. The `when` fill fire once when `isAuth` is false and then disposes itself.
The `reaction` will fire every time `isAuth` changes its state.

**Comment by leonardocustodio** [comment]
Closing the issue since the question was answered.
